cmd/autoretrieve/main.go

```diff
@@ -0,0 +1,775 @@
+		if r.config.minerBlacklist[query.Candidate.Miner] {
+			logger.Warnf("Skipping blacklisted miner %s for %s", query.Candidate.Miner, query.Candidate.RootCid)
+			continue
+		}
+
+		r.retrievalsInProgressLk.Lock()
+		logger.Infof(
+			"Attempting retrieval %v/%v from miner %s for %s (%v retrievals in progress)",
+			i+1,
+			len(queries),
+			query.Candidate.Miner,
+			query.Candidate.RootCid,
+			len(r.retrievalsInProgress),
+		)
+		r.retrievalsInProgressLk.Unlock()
+
+		var err error
+		stats, err = r.retrieve(ctx, query)
+		if err != nil {
+
+			logger.Errorf(
+				"Failed to retrieve %v/%v from miner %s for %s: %v",
+				i+1,
+				len(queries),
+				query.Candidate.Miner,
+				query.Candidate.RootCid,
+				err,
+			)
+
+			if errors.Is(err, ErrRetrievalAlreadyRunning) {
+				return err
+			}
+
+			continue
+		}
+
+		logger.Infof(
+			"Retrieval %v/%v succeeded from miner %s for %s\n\t"+
+				"Duration: %s\n\t"+
+				"Size: %s\n\t"+
+				"Average Speed: %s/s\n\t"+
+				"Total Payment: %s",
+			i+1,
+			len(queries),
+			query.Candidate.Miner,
+			query.Candidate.RootCid,
+			stats.Duration,
+			humanize.IBytes(stats.Size),
+			humanize.IBytes(stats.AverageSpeed),
+			types.FIL(stats.TotalPayment),
+		)
+
+		break
+	}
+
+	if stats == nil {
+		return xerrors.New("all retrievals failed")
+	}
+
+	return nil
+}
+
+var ErrRetrievalAlreadyRunning = xerrors.New("retrieval already running")
+
+func (r *bsnetReceiver) retrieve(ctx context.Context, query CandidateQuery) (*filclient.RetrievalStats, error) {
+	r.retrievalsInProgressLk.Lock()
+
+	// If we identify a retrieval already running for a potential root CID at
+	// this point, fail out immediately, this retrieval will be pointless
+	if _, ok := r.retrievalsInProgress[query.Candidate.RootCid]; ok {
+		r.retrievalsInProgressLk.Unlock()
+		return nil, fmt.Errorf("%w: root cid %s", ErrRetrievalAlreadyRunning, query.Candidate.RootCid)
+	}
+
+	// Mark the root CID's retrieval as in progress
+	r.retrievalsInProgress[query.Candidate.RootCid] = true
+
+	r.retrievalsInProgressLk.Unlock()
+
+	defer func() {
+		r.retrievalsInProgressLk.Lock()
+		delete(r.retrievalsInProgress, query.Candidate.RootCid)
+		r.retrievalsInProgressLk.Unlock()
+	}()
+
+	proposal, err := retrievehelper.RetrievalProposalForAsk(query.Response, query.Candidate.RootCid, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	retrieveCtx, retrieveCancel := context.WithCancel(ctx)
+	var lastBytesReceived uint64 = 0
+	var doneLk sync.Mutex
+	done := false
+	lastBytesReceivedTimer := time.AfterFunc(r.config.retrievalTimeout, func() {
+		doneLk.Lock()
+		done = true
+		doneLk.Unlock()
+
+		retrieveCancel()
+		logger.Errorf("Retrieval timed out after not receiving data for %s (stopped at %s)", r.config.retrievalTimeout, humanize.IBytes(lastBytesReceived))
+	})
+	stats, err := r.fc.RetrieveContentWithProgressCallback(retrieveCtx, query.Candidate.Miner, proposal, func(bytesReceived uint64) {
+		doneLk.Lock()
+		if !done {
```

Comment:
would it be less contentious to unlock immediately after retrieving the value of `done` here, instead of after the logic block, i.e.:

```go
doneLk.Lock()
isDone := done
doneLk.Unlock()
if !isDone {
  //...
}
```

Reply:
we also need to lock around `lastBytesReceived` which is used inside the if, and i believe we also need exclusive access to do the timer reset steps in order so its state doesn't get messed up. an instance of this mutex should only get locked and unlocked around 10 times throughout its lifetime of a few seconds, maybe 100 in rare cases. so i don't think it should be much of a problem.

Reply:
actually, if you have another idea on how to detect retrieval hangs, i'd be open to trying something else. using Timer was the quickest way i was able to figure it out, but it does seem pretty hacky